Semantic implementation section (#71)
chhoumann/p4: Report/sections/3_implementation/expression_type_checking.tex

\subsubsection*{Expression type checking} \label{sec:ExpressionTypeChecking}
To illustrate an example of how type checking is implemented in \dazel{}, we will now describe the \textbf{ExpressionTypeChecker}.
Note that we do not exclusively perform type checking on expressions, however this example sufficiently highlights our approach to type checking.

> Once the type checker visits an assignment node, it initializes an instance of the \textbf{ExpressionTypeChecker} whose responsibility is to traverse the expression and verify that the expression is legal by implementing the \textbf{IExpressionVisitor} interface. 

Reviewer: ```suggestion
Once the \textbf{TypeChecker} visits an assignment node, it initializes an instance of the \textbf{ExpressionTypeChecker}. The \textbf{ExpressionTypeChecker} is a visitor designed to traverse the nodes of an expression. While traversing, it verifies that the expression is legal.
```

Reviewer: I think this makes it clearer that the ExpressionTypeChecker is not _the_ type checker. I also think it improves the argument by not stating that it verifies that the expression is legal by 'implementing the \textbf{IExpressionVisitor} interface.'

> The \textbf{ExpressionTypeChecker} delegates the responsibility for storing the previously visited type to a class called \textbf{TypeHandler}. This class contains some simple rules that determine if types are compatible and can be seen on \snipref{lst:TypeHandler}. 

Reviewer: ```suggestion
The \textbf{ExpressionTypeChecker} delegates the responsibility for storing the previously visited type to a class called \textbf{TypeHandler}. This class contains some simple rules that determine if types are compatible and can be seen in \snipref{lst:TypeHandler}. 
```

Reviewer: Major nitpick

> Its rules allow floats and integers to occur in the same expression. As can be seen on lines \ref{line:AllowInt1} to \ref{line:AllowInt2}, if an expression is of type float and an integer is encountered, it remains a float expression.

Reviewer: ```suggestion
Its rules allow floats and integers to occur in the same expression. As can be seen on lines \ref{line:AllowInt1} to \ref{line:AllowInt2}. If an expression is of type float and an integer is encountered, it remains a float expression.
```

Major nit
